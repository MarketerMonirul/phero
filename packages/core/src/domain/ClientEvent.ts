export enum ClientEvent {
  ClientDidRegister = "CLIENT_DID_REGISTER",
  ManifestDidCHange = "MANIFEST_DID_CHANGE",
}
