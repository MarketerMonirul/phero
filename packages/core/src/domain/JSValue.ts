export type JSValue =
  | JSUntypedValue
  | JSStringValue
  | JSNumberValue
  | JSBooleanValue
  | JSObjectValue
  | JSArrayValue
  | JSDateValue
  | JSNullValue
  | JSUndefinedValue
  | JSOneOfTypesValue
  | JSTupleValue
  | JSModelRef

export enum JSType {
  untyped = "untyped",

  string = "string",
  number = "number",
  boolean = "boolean",

  object = "object",
  array = "array",
  date = "date",

  null = "null",
  undefined = "undefined",

  oneOfTypes = "oneOfTypes",
  tuple = "tuple",

  ref = "ref",
}

export interface JSUntypedValue {
  type: JSType.untyped
}

export interface JSStringValue {
  type: JSType.string
  oneOf?: string[]
}

export interface JSNumberValue {
  type: JSType.number
  oneOf?: number[]
}

export interface JSBooleanValue {
  type: JSType.boolean
  oneOf?: boolean[]
}

export type JSProperty = JSValue & { name: string }

export interface JSObjectValue {
  type: JSType.object
  properties: JSProperty[]
}

export interface JSArrayValue {
  type: JSType.array
  elementType: JSValue
}

export interface JSDateValue {
  type: JSType.date
}

export interface JSNullValue {
  type: JSType.null
}

export interface JSUndefinedValue {
  type: JSType.undefined
}

export interface JSOneOfTypesValue {
  type: JSType.oneOfTypes
  oneOfTypes: JSValue[]
}

export interface JSTupleValue {
  type: JSType.tuple
  elementTypes: JSValue[]
}

export interface JSModelRef {
  type: JSType.ref
  id: string
}
