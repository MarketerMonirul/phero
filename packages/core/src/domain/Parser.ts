export type Parser<T> = (data: unknown) => ParseResult<T>

export type ParseResult<T> = DataParseSuccess<T> | DataParseFailure

export interface DataParseSuccess<T> {
  ok: true
  result: T
}

export interface DataParseFailure {
  ok: false
  errors: DataParseError[]
}

export interface DataParseError {
  message: string
  path?: string
  errors?: DataParseError[]
}
