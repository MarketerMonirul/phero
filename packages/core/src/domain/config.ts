export interface SamenConfig {
  clients: string[]
}

export enum ClientEnvironment {
  Browser,
  Node,
}

export interface ClientConfig {
  development: { url: string }
  production: { url: string }
  env?: ClientEnvironment
}
