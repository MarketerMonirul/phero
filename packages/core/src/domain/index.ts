export * from "./manifest"
export * from "./JSValue"
export * from "./config"
export * from "./ClientEvent"
