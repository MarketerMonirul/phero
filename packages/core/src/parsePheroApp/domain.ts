import ts from "typescript"

export interface PheroApp {
  models: PheroModel[]
  errors: PheroError[]
  services: PheroService[]
}

export interface PheroModel {
  name: string
  ref: Model
}

export type Model =
  | ts.InterfaceDeclaration
  | ts.TypeAliasDeclaration
  | ts.EnumDeclaration

export interface PheroError {
  name: string
  sourceFile: string
  properties: PheroErrorProperty[]
  ref: ts.ClassDeclaration
}

export interface PheroErrorProperty {
  name: string
  type: ts.TypeNode
}

export interface PheroService {
  name: string
  funcs: PheroFunction[]
  config: PheroServiceConfig
  ref: ts.Node
}

export interface PheroFunction {
  name: string
  parameters: ts.ParameterDeclaration[]
  returnType: ts.TypeNode
  serviceContext?: {
    type: ts.TypeNode
    paramName?: string
  }
  ref: ts.FunctionLikeDeclarationBase
}

export interface PheroServiceConfig {
  middleware?: PheroMiddlewareConfig[]
  contextType?: ts.TypeNode
}

export interface PheroMiddlewareConfig {
  paramsType: ts.TypeNode
  nextType: ts.TypeNode | undefined
  contextType: ts.TypeNode
  middleware: ts.FunctionLikeDeclarationBase
}
