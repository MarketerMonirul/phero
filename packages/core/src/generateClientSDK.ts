import { RPCFunction, SamenManifest } from "./domain/manifest"
import { JSType, JSValue } from "./domain/JSValue"
import { formatCode } from "./utils"
import transformManifest from "./transformManifest"

export default function generateClientSDK(
  manifestPath: string,
  sdkPath: string,
): Promise<void> {
  return transformManifest(manifestPath, sdkPath, transformToClientSDK)
}

export function transformToClientSDK(manifest: SamenManifest): string {
  return formatCode(`
    import request from './request';

    ${Object.keys(manifest.models).map(
      (modelId) => `export ${manifest.models[modelId].ts}`,
    )}

    ${manifest.rpcFunctions.map(genRPC)}
  `)
}
const genRPC = (rpc: RPCFunction): string => {
  const { name } = rpc
  const params = rpc.parameters.map((p) => `${p.name}: ${genType(p.value)}`)
  const body = `{${rpc.parameters.map((p) => p.name).join(",")}}`
  const returnType = genType(rpc.returnType)
  const isVoid = rpc.returnType.type === JSType.untyped ? "true" : "false"

  return `
    export async function ${name}(${params}): Promise<${returnType}> {
      return request("${name}", ${body}, ${isVoid});
    }
  `
}

const genType = (value: JSValue): string => {
  switch (value.type) {
    case JSType.number:
    case JSType.string:
    case JSType.boolean:
    case JSType.null:
    case JSType.undefined:
      return value.type

    case JSType.ref:
      return value.id

    case JSType.array:
      return `${value.elementType.type}[]`

    case JSType.tuple:
      return `[${value.elementTypes.map(genType).join(", ")}]`

    case JSType.date:
      return "Date"

    case JSType.oneOfTypes:
      return `${value.oneOfTypes.map(genType).join(" | ")}`

    case JSType.object:
      return `{${value.properties
        .map((p) => `${p.name}: ${genType(p)}`)
        .join(";")}}`

    case JSType.untyped:
      return "void"
  }
}
