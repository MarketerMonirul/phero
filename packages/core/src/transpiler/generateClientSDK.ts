import { promises as fs } from "fs"
import { Project } from "ts-morph"
import { JSType } from "../domain/JSValue"
import { RPCFunction, SamenManifest } from "../domain/manifest"
import { ClientSDKCompilerError, validateProject } from "../errors"
import * as paths from "../paths"
import { formatCode, generateParameters, generateType } from "./utils"

export default async function generateClientSDK(
  projectDir: string,
): Promise<void> {
  try {
    const project = new Project({
      compilerOptions: {
        declaration: true,
        outDir: paths.clientSdkDir(projectDir),
      },
    })
    const manifest = JSON.parse(
      await fs.readFile(paths.clientManifestFile(projectDir), "utf-8"),
    )
    project.createSourceFile("index.ts", transformToClientSDK(manifest))
    validateProject(project)
    await project.emit()
  } catch (error) {
    throw new ClientSDKCompilerError(error)
  }
}

export function transformToClientSDK(manifest: SamenManifest): string {
  return formatCode(`
    ${requestFunction}

    ${Object.keys(manifest.models).map(
      (modelId) => `export ${manifest.models[modelId].ts}`,
    )}

    ${manifest.rpcFunctions.map(genRPC).join("\n")}
  `)
}

const requestFunction = `
  // TODO: Get from manifest
  const ENDPOINT = "http://localhost:4000/"

  async function request<T>(
    name: string,
    body: object,
    isVoid: boolean,
  ): Promise<T> {
    try {
      const result = await fetch(ENDPOINT + name, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      })
      if (!result.ok) {
        throw new Error(\`Call failed with status \${result.status}\`)
      }
      if (!isVoid) {
        const data = await result.json()
        return data as T
      }
    } catch (err) {
      console.error(err)
      throw new Error("Network error")
    }
  }
`

const genRPC = (rpc: RPCFunction): string => {
  const { name } = rpc
  const params = generateParameters(rpc.parameters)
  const returnType = `Promise<${generateType(rpc.returnType)}>`
  const body = `{${rpc.parameters.map((p) => p.name).join(",")}}`
  const isVoid = rpc.returnType.type === JSType.untyped ? "true" : "false"

  return `
    export async function ${name}(${params}): ${returnType} {
      return request("${name}", ${body}, ${isVoid});
    }
  `
}
