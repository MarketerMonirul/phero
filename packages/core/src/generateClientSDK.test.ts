import generateClientSDK, { transformToClientSDK } from "./generateClientSDK"
import { RPCFunctionParameter } from "./domain/manifest"
import { JSType, JSValue } from "./domain/JSValue"

describe("rpcFunctions", () => {
  describe("return type", () => {
    const gen = (returnType: JSValue): string =>
      transformToClientSDK({
        models: {},
        refs: {},
        rpcFunctions: [
          {
            name: "something",
            parameters: [],
            returnType,
            modelIds: [],
          },
        ],
      })

    test("string", () => {
      expect(gen({ type: JSType.string })).toMatchSnapshot()
    })

    test("number", () => {
      expect(gen({ type: JSType.number })).toMatchSnapshot()
    })

    test("boolean", () => {
      expect(gen({ type: JSType.boolean })).toMatchSnapshot()
    })

    test("object", () => {
      expect(
        gen({
          type: JSType.object,
          properties: [
            { name: "a", type: JSType.boolean },
            { name: "b", type: JSType.string },
          ],
        }),
      ).toMatchSnapshot()
    })

    test("array", () => {
      expect(
        gen({ type: JSType.array, elementType: { type: JSType.number } }),
      ).toMatchSnapshot()
    })

    test("date", () => {
      expect(gen({ type: JSType.date })).toMatchSnapshot()
    })

    test("null", () => {
      expect(gen({ type: JSType.null })).toMatchSnapshot()
    })

    test("undefined", () => {
      expect(gen({ type: JSType.undefined })).toMatchSnapshot()
    })

    test("oneOfTypes", () => {
      expect(
        gen({
          type: JSType.oneOfTypes,
          oneOfTypes: [{ type: JSType.boolean }, { type: JSType.string }],
        }),
      ).toMatchSnapshot()
    })

    test("tuple", () => {
      expect(
        gen({
          type: JSType.tuple,
          elementTypes: [
            { type: JSType.date },
            { type: JSType.array, elementType: { type: JSType.number } },
          ],
        }),
      ).toMatchSnapshot()
    })

    test("ref", () => {
      const code = transformToClientSDK({
        models: {
          Something: {
            id: "Something",
            ts: "interface Something { x: boolean }",
          },
        },
        refs: {
          Something: {
            id: "Something",
            modelId: "Something",
            value: {
              type: JSType.object,
              properties: [
                {
                  name: "x",
                  type: JSType.boolean,
                },
              ],
            },
          },
        },
        rpcFunctions: [
          {
            name: "something",
            parameters: [],
            returnType: { type: JSType.ref, id: "Something" },
            modelIds: [],
          },
        ],
      })

      expect(code).toMatchSnapshot()
    })

    test("untyped", () => {
      expect(gen({ type: JSType.untyped })).toMatchSnapshot()
    })
  })

  describe("parameters", () => {
    const gen = (parameters: RPCFunctionParameter[]): string =>
      transformToClientSDK({
        rpcFunctions: [
          {
            name: "something",
            parameters,
            returnType: { type: JSType.untyped },
            modelIds: [],
          },
        ],
        models: {},
        refs: {},
      })

    test("string", () => {
      expect(
        gen([{ name: "a", index: 0, value: { type: JSType.string } }]),
      ).toMatchSnapshot()
    })

    test("number", () => {
      expect(
        gen([{ name: "a", index: 0, value: { type: JSType.number } }]),
      ).toMatchSnapshot()
    })

    test("boolean", () => {
      expect(
        gen([{ name: "a", index: 0, value: { type: JSType.boolean } }]),
      ).toMatchSnapshot()
    })

    test("object", () => {
      expect(
        gen([
          {
            name: "a",
            index: 0,
            value: {
              type: JSType.object,
              properties: [{ name: "a", type: JSType.number }],
            },
          },
        ]),
      ).toMatchSnapshot()
    })

    test("array", () => {
      expect(
        gen([
          {
            name: "a",
            index: 0,
            value: { type: JSType.array, elementType: { type: JSType.number } },
          },
        ]),
      ).toMatchSnapshot()
    })

    test("date", () => {
      expect(
        gen([{ name: "a", index: 0, value: { type: JSType.date } }]),
      ).toMatchSnapshot()
    })

    test("null", () => {
      expect(
        gen([{ name: "a", index: 0, value: { type: JSType.null } }]),
      ).toMatchSnapshot()
    })

    test("undefined", () => {
      expect(
        gen([{ name: "a", index: 0, value: { type: JSType.undefined } }]),
      ).toMatchSnapshot()
    })

    test("oneOfTypes", () => {
      expect(
        gen([
          {
            name: "a",
            index: 0,
            value: {
              type: JSType.oneOfTypes,
              oneOfTypes: [{ type: JSType.boolean }, { type: JSType.number }],
            },
          },
        ]),
      ).toMatchSnapshot()
    })

    test("tuple", () => {
      expect(
        gen([
          {
            name: "a",
            index: 0,
            value: {
              type: JSType.tuple,
              elementTypes: [{ type: JSType.date }, { type: JSType.boolean }],
            },
          },
        ]),
      ).toMatchSnapshot()
    })

    test("ref", () => {
      const code = transformToClientSDK({
        models: {
          Something: {
            id: "Something",
            ts: "interface Something { x: boolean }",
          },
        },
        refs: {
          Something: {
            id: "Something",
            modelId: "Something",
            value: {
              type: JSType.object,
              properties: [
                {
                  name: "x",
                  type: JSType.boolean,
                },
              ],
            },
          },
        },
        rpcFunctions: [
          {
            name: "something",
            parameters: [
              {
                index: 0,
                name: "a",
                value: {
                  type: JSType.ref,
                  id: "Something",
                },
              },
            ],
            returnType: { type: JSType.untyped },
            modelIds: ["Something"],
          },
        ],
      })
      expect(code).toMatchSnapshot()
    })

    test("no parameters", () => {
      expect(gen([])).toMatchSnapshot()
    })

    test("multiple parameters", () => {
      expect(
        gen([
          { name: "a", index: 0, value: { type: JSType.number } },
          { name: "b", index: 1, value: { type: JSType.string } },
        ]),
      ).toMatchSnapshot()
    })
  })
})
