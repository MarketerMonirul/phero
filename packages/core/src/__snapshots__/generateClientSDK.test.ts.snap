// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rpcFunctions parameters array 1`] = `
"import request from \\"./request\\";

export async function something(a: number[]): Promise<void> {
  return request(\\"something\\", { a }, true);
}
"
`;

exports[`rpcFunctions parameters boolean 1`] = `
"import request from \\"./request\\";

export async function something(a: boolean): Promise<void> {
  return request(\\"something\\", { a }, true);
}
"
`;

exports[`rpcFunctions parameters date 1`] = `
"import request from \\"./request\\";

export async function something(a: Date): Promise<void> {
  return request(\\"something\\", { a }, true);
}
"
`;

exports[`rpcFunctions parameters multiple parameters 1`] = `
"import request from \\"./request\\";

export async function something(a: number, b: string): Promise<void> {
  return request(\\"something\\", { a, b }, true);
}
"
`;

exports[`rpcFunctions parameters no parameters 1`] = `
"import request from \\"./request\\";

export async function something(): Promise<void> {
  return request(\\"something\\", {}, true);
}
"
`;

exports[`rpcFunctions parameters null 1`] = `
"import request from \\"./request\\";

export async function something(a: null): Promise<void> {
  return request(\\"something\\", { a }, true);
}
"
`;

exports[`rpcFunctions parameters number 1`] = `
"import request from \\"./request\\";

export async function something(a: number): Promise<void> {
  return request(\\"something\\", { a }, true);
}
"
`;

exports[`rpcFunctions parameters object 1`] = `
"import request from \\"./request\\";

export async function something(a: { a: number }): Promise<void> {
  return request(\\"something\\", { a }, true);
}
"
`;

exports[`rpcFunctions parameters oneOfTypes 1`] = `
"import request from \\"./request\\";

export async function something(a: boolean | number): Promise<void> {
  return request(\\"something\\", { a }, true);
}
"
`;

exports[`rpcFunctions parameters ref 1`] = `
"import request from \\"./request\\";

export interface Something {
  x: boolean;
}

export async function something(a: Something): Promise<void> {
  return request(\\"something\\", { a }, true);
}
"
`;

exports[`rpcFunctions parameters string 1`] = `
"import request from \\"./request\\";

export async function something(a: string): Promise<void> {
  return request(\\"something\\", { a }, true);
}
"
`;

exports[`rpcFunctions parameters tuple 1`] = `
"import request from \\"./request\\";

export async function something(a: [Date, boolean]): Promise<void> {
  return request(\\"something\\", { a }, true);
}
"
`;

exports[`rpcFunctions parameters undefined 1`] = `
"import request from \\"./request\\";

export async function something(a: undefined): Promise<void> {
  return request(\\"something\\", { a }, true);
}
"
`;

exports[`rpcFunctions return type array 1`] = `
"import request from \\"./request\\";

export async function something(): Promise<number[]> {
  return request(\\"something\\", {}, false);
}
"
`;

exports[`rpcFunctions return type boolean 1`] = `
"import request from \\"./request\\";

export async function something(): Promise<boolean> {
  return request(\\"something\\", {}, false);
}
"
`;

exports[`rpcFunctions return type date 1`] = `
"import request from \\"./request\\";

export async function something(): Promise<Date> {
  return request(\\"something\\", {}, false);
}
"
`;

exports[`rpcFunctions return type null 1`] = `
"import request from \\"./request\\";

export async function something(): Promise<null> {
  return request(\\"something\\", {}, false);
}
"
`;

exports[`rpcFunctions return type number 1`] = `
"import request from \\"./request\\";

export async function something(): Promise<number> {
  return request(\\"something\\", {}, false);
}
"
`;

exports[`rpcFunctions return type object 1`] = `
"import request from \\"./request\\";

export async function something(): Promise<{ a: boolean; b: string }> {
  return request(\\"something\\", {}, false);
}
"
`;

exports[`rpcFunctions return type oneOfTypes 1`] = `
"import request from \\"./request\\";

export async function something(): Promise<boolean | string> {
  return request(\\"something\\", {}, false);
}
"
`;

exports[`rpcFunctions return type ref 1`] = `
"import request from \\"./request\\";

export interface Something {
  x: boolean;
}

export async function something(): Promise<Something> {
  return request(\\"something\\", {}, false);
}
"
`;

exports[`rpcFunctions return type string 1`] = `
"import request from \\"./request\\";

export async function something(): Promise<string> {
  return request(\\"something\\", {}, false);
}
"
`;

exports[`rpcFunctions return type tuple 1`] = `
"import request from \\"./request\\";

export async function something(): Promise<[Date, number[]]> {
  return request(\\"something\\", {}, false);
}
"
`;

exports[`rpcFunctions return type undefined 1`] = `
"import request from \\"./request\\";

export async function something(): Promise<undefined> {
  return request(\\"something\\", {}, false);
}
"
`;

exports[`rpcFunctions return type untyped 1`] = `
"import request from \\"./request\\";

export async function something(): Promise<void> {
  return request(\\"something\\", {}, true);
}
"
`;
